File: ipadOS/RCForb/Sources/Services/AuthService.swift
import Foundation

enum AuthService {
    private static let loginURL = "https://api.remotehams.com/v2/login.php"

    static func authenticate(_ user: String, _ password: String) async -> AuthResult {
        let passMD5 = md5(password)
        return await authenticateWithMD5(user, passMD5)
    }

    static func authenticateWithMD5(_ user: String, _ passwordMD5: String) async -> AuthResult {
        let passDoubleHash = md5(passwordMD5)
        let encodedUser = user.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? user
        let valid = validationToken(user, passDoubleHash)

        let body = "user=\(encodedUser)&pass=\(passDoubleHash)&valid=\(valid)&getkey=true"

        guard let url = URL(string: loginURL) else {
            return AuthResult(success: false, message: "Invalid URL")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = body.data(using: .utf8)

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            let text = String(data: data, encoding: .utf8) ?? ""

            if text.hasPrefix("Valid") {
                let parts = text.components(separatedBy: ",")
                let apiKey = parts.count > 1 ? parts[1].trimmingCharacters(in: .whitespaces) : nil
                return AuthResult(success: true, message: text, apiKey: apiKey)
            }
            return AuthResult(success: false, message: text)
        } catch {
            return AuthResult(success: false, message: error.localizedDescription)
        }
    }

    static func trackOnline(_ user: String, _ passwordMD5: String, _ orbId: String?) async -> Bool {
        let passDoubleHash = md5(passwordMD5)
        let encodedUser = user.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? user
        let valid = md5(encodedUser + passDoubleHash)

        var body = "user=\(encodedUser)&pass=\(passDoubleHash)&varMe=valYou&logonline=true&valid=\(valid)"
        if let orbId { body += "&orbid=\(orbId)" }

        guard let url = URL(string: loginURL) else { return false }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = body.data(using: .utf8)

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            let text = String(data: data, encoding: .utf8) ?? ""
            return text.contains("Valid")
        } catch {
            return false
        }
    }
}

File: android/app/src/main/java/com/rcforb/android/services/AuthService.kt
package com.rcforb.android.services

import com.rcforb.android.models.AuthResult
import com.rcforb.android.protocol.md5
import com.rcforb.android.protocol.validationToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.FormBody
import okhttp3.OkHttpClient
import okhttp3.Request
import java.net.URLEncoder

object AuthService {
    private const val LOGIN_URL = "https://api.remotehams.com/v2/login.php"
    private val client = OkHttpClient()

    suspend fun authenticate(user: String, password: String): AuthResult {
        val passMD5 = md5(password)
        return authenticateWithMD5(user, passMD5)
    }

    suspend fun authenticateWithMD5(user: String, passwordMD5: String): AuthResult = withContext(Dispatchers.IO) {
        try {
            val passDoubleHash = md5(passwordMD5)
            val encodedUser = URLEncoder.encode(user, "UTF-8")
            val valid = validationToken(user, passDoubleHash)

            val body = FormBody.Builder()
                .addEncoded("user", encodedUser)
                .addEncoded("pass", passDoubleHash)
                .addEncoded("valid", valid)
                .addEncoded("getkey", "true")
                .build()

            val request = Request.Builder()
                .url(LOGIN_URL)
                .post(body)
                .build()

            val response = client.newCall(request).execute()
            val text = response.body?.string() ?: ""

            if (text.startsWith("Valid")) {
                val parts = text.split(",")
                val apiKey = if (parts.size > 1) parts[1].trim() else null
                AuthResult(success = true, message = text, apiKey = apiKey)
            } else {
                AuthResult(success = false, message = text)
            }
        } catch (e: Exception) {
            AuthResult(success = false, message = e.localizedMessage ?: "Network error")
        }
    }

    suspend fun trackOnline(user: String, passwordMD5: String, orbId: String?): Boolean = withContext(Dispatchers.IO) {
        try {
            val passDoubleHash = md5(passwordMD5)
            val encodedUser = URLEncoder.encode(user, "UTF-8")
            val valid = md5(encodedUser + passDoubleHash)

            val bodyBuilder = FormBody.Builder()
                .addEncoded("user", encodedUser)
                .addEncoded("pass", passDoubleHash)
                .addEncoded("varMe", "valYou")
                .addEncoded("logonline", "true")
                .addEncoded("valid", valid)
            if (orbId != null) bodyBuilder.addEncoded("orbid", orbId)

            val request = Request.Builder()
                .url(LOGIN_URL)
                .post(bodyBuilder.build())
                .build()

            val response = client.newCall(request).execute()
            val text = response.body?.string() ?: ""
            text.contains("Valid")
        } catch (_: Exception) {
            false
        }
    }
}
